source/game_sa/CBuoyancy.cpp

```diff
@@ -64,49 +77,475 @@ bool cBuoyancy::ProcessBuoyancy(CPhysical* pEntity, float fBuoyancy, CVector* pV
+            m_vecBoundingMax.y *= 1.4F;
+            break;
+        default:
+            if (bIsVehicle && ((CVehicle*)pEntity)->m_nVehicleSubClass == 3) {
+                m_vecBoundingMin.y = -m_vecBoundingMax.y;
+                m_vecBoundingMax.z = m_vecBoundingMin.z * -1.1F;
+                m_vecBoundingMin.z *= 0.85F;
+            }
+            break;
+        }
+    }
+    else {
+        switch (pEntity->m_nModelIndex) {
+        case eModelID::MODEL_SQUALO: //446
+            m_vecBoundingMax.y *= 0.9F;
+            m_vecBoundingMin.y *= 0.9F;
+            break;
+        case eModelID::MODEL_SPEEDER: //452
+            m_vecBoundingMax.y *= 1.25F;
+            m_vecBoundingMin.y *= 0.83F;
+            break;
+        case eModelID::MODEL_REEFER: //453
+        case eModelID::MODEL_JETMAX: //498
+            m_vecBoundingMin.y *= 0.9F;
+            break;
+        case eModelID::MODEL_TROPIC: //454
+            m_vecBoundingMax.y *= 1.3F;
+            m_vecBoundingMin.y *= 0.82F;
+            m_vecBoundingMin.z -= 0.2F;
+            break;
+        case eModelID::MODEL_COASTG: //472
+            m_vecBoundingMax.y *= 1.1F;
+            m_vecBoundingMin.y *= 0.9F;
+            m_vecBoundingMin.z -= 0.3F;
+            break;
+        case eModelID::MODEL_DINGHY: //473
+            m_vecBoundingMax.y *= 1.3F;
+            m_vecBoundingMin.y *= 0.9F;
+            m_vecBoundingMin.z -= 0.2F;
+            break;
+        case eModelID::MODEL_MARQUIS: //484
+            m_vecBoundingMax.y *= 1.1F;
+            m_vecBoundingMin.y *= 0.9F;
+            break;
+        case eModelID::MODEL_LAUNCH: //595
+            m_vecBoundingMax.y *= 1.25F;
+            m_vecBoundingMin.y *= 0.8F;
+            m_vecBoundingMin.z -= 0.1F;
+            break;
+        default:
+            m_vecBoundingMax.y *= 1.05F;
+            m_vecBoundingMin.y *= 0.9F;
+            break;
+        }
+    }
+
+    // Calculate center offset and "normalized" version of it
+    // (it's not normalised vector, just the offset scaled by multiplicative inverse of it's longest component
+    //  so the longest component is equal to 1.0F)
+    m_vecCenterOffset = (m_vecBoundingMax - m_vecBoundingMin) * 0.5F;
+    float fScale;
+    if ((m_vecCenterOffset.z > m_vecCenterOffset.x) && (m_vecCenterOffset.z > m_vecCenterOffset.y))
+        fScale = 1.0F / m_vecCenterOffset.z;
+    else if ((m_vecCenterOffset.y > m_vecCenterOffset.x) && (m_vecCenterOffset.y > m_vecCenterOffset.z))
+        fScale = 1.0F / m_vecCenterOffset.y;
+    else
+        fScale = 1.0F / m_vecCenterOffset.x;
+
+    m_vecNormalizedCenterOffset = m_vecCenterOffset * fScale;
+    m_fNumCheckedPoints = 1.0F;
+    m_vecMoveForce.Set(0.0F, 0.0F, 0.0F);
+    m_bInWater = false;
+    m_fEntityWaterImmersion = 0.0F;
+    m_vecPos = pEntity->GetPosition();
+    m_fWaterLevel += m_fUnkn2;
+
+    m_unknStruct.iUnkn1 = 0;
+    m_unknStruct.iUnkn2 = 0;
+    m_unknStruct.fEntityPosZ = m_vecPos.z;
```

Comment:
```suggestion
    m_savedPosition = CVector(0.0f, 0.0f, m_vecPos.z);
```
This is actually CVector.

Reply:
Changed it to vector